Advanced Transformations (Public Beta) (#4302)
segmentio/segment-docs: src/protocols/transform/index.md

Transformations can be deleted and edited by clicking on the overflow menu. When editing a Transformation, only the resulting event or property names, and Transformation name can be edited. If you want to select a different event or source, create a separate Transformation rule.

> note "Transformations created using the Public API"

> > On the Transformations page in the Segment app, you can view and rename transformations that you created with the Public API. You can only edit these transformations using the Public API.

Reviewer: I'm not sure if this is true for all cases. We should try for simplest case and see if UI allows edit.

Transformations can be deleted and edited by clicking on the overflow menu. When editing a Transformation, only the resulting event or property names, and Transformation name can be edited. If you want to select a different event or source, create a separate Transformation rule.

> note "Transformations created using the Public API"

> > On the Transformations page in the Segment app, you can view and rename transformations that you created with the Public API. In some cases, you are allowed to edit these transformations in the UI.

Author: ```suggesting

```

> > On the Transformations page in the Segment app, you can view and rename transformations that you created with the Public API. In some cases, you are allowed to edit these transformations in the UI.

Author: ```suggestion
On the Transformations page in the Segment app, you can view and rename transformations that you created with the Public API. In some cases, you can edit these transformations in the UI.
```

  - **Casing functions:** Use [Segment's Public API](https://docs.segmentapis.com/tag/Transformations){:target="_blank"} to transform property value casing to `lowercase`, `uppercase`, `snakecase`, `kebabcase`, or `titlecase`. When transforming data with casing functions, use the [`fqlDefinedProperties`](https://docs.segmentapis.com/tag/Transformations#operation/createTransformation!ct=application/vnd.segment.v1+json&path=fqlDefinedProperties&t=request){:target="_blank"} array to define the FQL you want to use and the new or existing `propertyName` you'd like to transform.

>     - **Static and dynamic value casing:** You can use casing functions to transform property values to create uniform tracking data. For example, you can convert `united states` to `USA` to keep your downstream data consistent. <br/>You can transform these properties using static casing functions: <br/>

Reviewer 2: ```suggestion
    - **Static and dynamic value casing:** Use casing functions to transform property values to create uniform tracking data. For example, you can convert `united states` to `USA` to keep your downstream data consistent. <br/>You can transform these properties using static casing functions: <br/>
```

- **Add a new property name and assign a value:** If you want to create a new property and set a static value, use [Segment's Public API](https://docs.segmentapis.com/tag/Transformations){:target="_blank"} to create `new_property: static_value`. Segment currently supports setting static values for top-level fields with `propertyValueTransformations`. However, Segment doesn't support changing fields outside the properties or traits object with `propertyRenames`.

> - **Change property value casing:** Use [Segment's Public API](https://docs.segmentapis.com/tag/Transformations){:target="_blank"} to transform property value casing to lowercase, uppercase, snake case, kebab case, or title case. For example, transform the property value `united states` to `USA` to remain consistent with your data tracking. You can transform these properties using static casing functions (`"uppercase("United States")"`) or dynamic casing functions (`"lowercase(property value 1")"`).
> 
> - **Create a new property**: Use [Segment's Public API](https://docs.segmentapis.com/tag/Transformations){:target="_blank"} to create a new property and set the value of the new property to the transformed value of an existing property. For example, create a new property (`prop2`) with a value of `lowercase(properties.prop1)` by including `prop2: lowercase(properties.prop1)` in your payload.
> 
> - **Replace the value of a current property**: You can replace the property value of an existing property using static casing functions, for example `prop1: "titlecase(value 2)"`, or dynamic casing functions (`"prop1: kebabcase("properties.prop1")"`) with [Segment's Public API](https://docs.segmentapis.com/tag/Transformations){:target="_blank"}.
> 
> - **Change the value of a property if a user triggers an event:** Use [Segment's Public API](https://docs.segmentapis.com/tag/Transformations){:target="_blank"} to dynamically assign the value of one existing property to another property, or assign the value of an property to a new property. For example, you can assign a new property, `property2`, to have the same value as `property1` if a user clicks a button. 

Reviewer: ```suggestion
- ** Property Transformations:**
  - **Assigning static values:**
    - If you want to create a new property and set a static value, use [Segment's Public API](<https://docs.segmentapis.com/tag/Transformations>){:target="_blank"} to create `new_property: static_value`. Segment currently supports setting static values for top-level fields, as well as fields within the `context` or `properties` object with `propertyValueTransformations`. However, Segment doesn't support changing fields outside the properties or traits object with `propertyRenames`.
    - You can use assign `propertyValueTransformations` a single object to target different fields to assign the same value or multiple objects to target differnt fields with their own static values.
  - **Casing functions:** Use [Segment's Public API](<https://docs.segmentapis.com/tag/Transformations>){:target="_blank"} to transform property value casing to `lowercase, uppercase, snakecase, kebabcase, or titlecase`. You can use `fqlDefinedProperties` to define the `fql` (with or without casing functions) you want to use to and the `propertyName` to apply it towards (this can be an existing property name or a new one).
    - **Static and Dynamic Value Casing:** For example, transform the property value `united states` to `USA` to remain consistent with your data tracking. You can transform these properties using static casing functions (`fqlDefinedProperties": [{"fql": "uppercasecase("United States)", "propertyName": "properties.propertyValue1"}]`) or dynamic casing functions (`fqlDefinedProperties": [{"fql": "lowercase(properties.propertyValue1)", "propertyName": "properties.propertyValue1"}]`).
    - **Create a new property with casing applied value**: Use [Segment's Public API](<https://docs.segmentapis.com/tag/Transformations>){:target="_blank"} to create a new property and set the value of the new property to the transformed value of an existing property. For example, create a new property (`prop2`) with a value of `lowercase(properties.prop1)` by including `fqlDefinedProperties": [{"fql": "lowercase(properties.prop1)", "propertyName": "properties.prop2"}]` in your payload.
      - You can dynamically assign the value of one existing property to another property, or assign the value of an property to a new property without applying casing functions.
  - Note that you can only assign one property to `fqlDefinedProperties` array.
```

- **Add a new property name and assign a value:** If you want to create a new property and set a static value, use [Segment's Public API](https://docs.segmentapis.com/tag/Transformations){:target="_blank"} to create `new_property: static_value`. Segment currently supports setting static values for top-level fields with `propertyValueTransformations`. However, Segment doesn't support changing fields outside the properties or traits object with `propertyRenames`.

> - **Change property value casing:** Transform property value casing to lowercase, uppercase, snake case, kebab case, or title case. For example, transform the property value `united states` to `USA` to remain consistent with your data tracking.

Reviewer: So, you can achieve these:

- static casing functions (upper/lower/title) – ex. lower(United States)
- dynamic casing (upper/lower/title) – ex. lower(property value 1)
  - above 2  should work for `kebabcase`, `titlecase`, `snakecase`, `lowercase`, `uppercase` 
  - also, this allows you to 
    - create a completely new property `prop2: lowercase(properties.prop1)` or 
    - replace value of the current one `prop1: kebabcase(properties.prop1) -- this is dynamic
    - replace value of the current one `prop1: camelcase("united states of america")` -- this is static
- dynamic property value changes – ex. If event = example event, prop 1 = prop 3
  - you can assign another property's value to the target property (either existing or create a new one)
